libdnf5/transaction/transaction_replay.cpp

```diff
@@ -0,0 +1,425 @@
+                raise_or_warn(skip_unavailable, fmt::format("Cannot find group with id: {:s}", group_id));
+                continue;
+            }
+            if (group_query.size() > 1) {
+                raise_or_warn(skip_unavailable, fmt::format("Multiple environments matching id: {:s}", group_id));
+                continue;
+            }
+            group_actions.push_back(
+                {action, transaction_item_reason_from_string(reason), group_query.get(), group_package_types});
+        }
+    }
+
+    // PARSE AND FIND PACKAGES
+
+    struct json_object * json_rpms = nullptr;
+    if (json_object_object_get_ex(data, "rpms", &json_rpms) != 0) {
+        std::string action;
+        std::string nevra;
+        TransactionItemReason reason = TransactionItemReason::NONE;
+        std::string group;
+        std::string repo_id;
+        auto len = json_object_array_length(json_rpms);
+
+        libdnf5::rpm::PackageQuery installed(base);
+        installed.filter_installed();
+
+        for (std::size_t i = 0; i < len; ++i) {
+            struct json_object * rpm = json_object_array_get_idx(json_rpms, i);
+            if (json_object_object_get_ex(rpm, "action", &value) != 0) {
+                action = json_object_get_string(value);
+            }
+            if (json_object_object_get_ex(rpm, "nevra", &value) != 0) {
+                nevra = json_object_get_string(value);
+            }
+            if (json_object_object_get_ex(rpm, "reason", &value) != 0) {
+                reason = transaction_item_reason_from_string(json_object_get_string(value));
+            }
+            if (json_object_object_get_ex(rpm, "group", &value) != 0) {
+                group = json_object_get_string(value);
+            }
+            if (json_object_object_get_ex(rpm, "repo_id", &value) != 0) {
+                repo_id = json_object_get_string(value);
+            }
+
+
+            // Find the required package
+            libdnf5::rpm::PackageQuery pkg_query(base);
+
+            // In case the package is found in the same repo as in the original transaction, limit the query to that
+            // plus installed packages. IOW remove packages with the same NEVRA in case they are found in multiple
+            // repos and the repo the package came from originally is one of them.
+            // This can e.g. make a difference in the system-upgrade plugin, in case the same NEVRA is in two repos,
+            // this makes sure the same repo is used for both download and upgrade steps of the plugin.
+            if (!repo_id.empty()) {
+                if (repo_id == "@System") {
+                    pkg_query = installed;
+                } else {
+                    libdnf5::rpm::PackageQuery query_repo(base);
+                    query_repo.filter_repo_id({repo_id});
+                    if (!query_repo.empty()) {
+                        pkg_query = query_repo;
+                        pkg_query |= installed;
+                    }
```

Comment:
This is interesting. If repository does not exist then it continues with all packages. May I ask you for an explanation why this behavior is required?

Reply:
I have taken this behavior from the dnf4 replay however I would be fine if we make it more strict and don't allow running without the repo specified.